pkg/plugins/backendplugin/instrumentation/instrumentation.go

```diff
@@ -58,29 +58,105 @@ const (
 	endpointQueryData      = "queryData"
 )
 
+type errorSource string
+
+const (
+	grafanaSource          errorSource = "grafana"
+	userSource             errorSource = "user"
+	downstreamClientSource errorSource = "downstreamClient"
+	downstreamServerSource errorSource = "downstreamServer"
+	noneSource             errorSource = "none"
+)
+
 var logger = plog.New("plugin.instrumentation")
 
 // instrumentPluginRequest instruments success rate and latency of `fn`
 func instrumentPluginRequest(ctx context.Context, cfg Cfg, pluginCtx *backend.PluginContext, endpoint string, fn func() error) error {
-	status := statusOK
-
 	start := time.Now()
 	timeBeforePluginRequest := log.TimeSinceStart(ctx, start)
 
 	err := fn()
-	if err != nil {
-		status = statusError
-		if errors.Is(err, context.Canceled) {
-			status = statusCancelled
+
+	elapsed := time.Since(start)
+	status := getStatus(err)
+
+	logger.Info("Plugin Request Completed")
```

Comment:
> Now handled in `logDatasourceRequests`, isn't it?

Correct

No this PR is not an implementation of the proposal in the document. What I am trying to achieve here is "similar" in regard to the end goal, which is to have more information about where an error come from, without breaking the way errors are currently dealt with. But instead of incoming http request, here it's plugin request, so a go function call.
 
I don't see a dependency between this PR and the work on the HTTP request, but maybe I am wrong or you would prefere to have aligned implementations or error source for examples

Reply:
I think the dependency might be how you decide/propagate the source of an error. I think the general idea regarding /api/ds/query is
- If at least one of the query data responses contains an error the HTTP status returned is 400 Bad Request and errorSource=downstream
- If an error is returned from a plugin/QueryData the HTTP status returned is 500 Internal Server Error and errorSource=downstream if the error originating from plugin and not Grafana and the question/similarity with your PR vs our issue is how do we know that

Does this make more sense?

Reply:
Yes thanks

Reply:
The log statement `logger.Info("Plugin Request Completed")` is already in `logDatasourceRequest` which you call further down in this function so is it really needed here as well? If so, why?
https://github.com/grafana/grafana/blob/6d8eac38bdf7286398b62cf7b25b68e49d1062d1/pkg/plugins/backendplugin/instrumentation/instrumentation.go#L85
https://github.com/grafana/grafana/blob/6d8eac38bdf7286398b62cf7b25b68e49d1062d1/pkg/plugins/backendplugin/instrumentation/instrumentation.go#L210